src/vdom.rs

```diff
@@ -371,50 +374,36 @@ impl VdomWeak {
     /// Replace the root rendering component with the new `root`.
     ///
     /// Returns a future that resolves to the *old* root component.
-    pub fn set_component(
+    pub async fn set_component(
         self,
         root: Box<dyn RootRender>,
-    ) -> impl Future<Item = Box<dyn RootRender + 'static>, Error = VdomDroppedError> {
-        futures::future::ok(self.inner.upgrade())
-            .and_then(|inner| inner.ok_or(()))
-            .map_err(|_| VdomDroppedError {})
-            .and_then(|inner| {
-                let promise = js_sys::Promise::resolve(&JsValue::null());
-                JsFuture::from(promise)
-                    .map_err(|_| VdomDroppedError {})
-                    .and_then(move |_| {
-                        let old = {
-                            let mut exclusive = inner.exclusive.borrow_mut();
-                            mem::replace(&mut *exclusive.component.as_mut().unwrap_throw(), root)
-                        };
-                        VdomWeak::new(&inner).render().map(|_| old)
-                    })
-            })
+    ) -> Result<Box<dyn RootRender + 'static>, VdomDroppedError> {
+        let inner = self.inner.upgrade().ok_or(VdomDroppedError {})?;
+
+        let old = {
+            let mut exclusive = inner.exclusive.borrow_mut();
+            mem::replace(&mut *exclusive.component.as_mut().unwrap_throw(), root)
+        };
+
+        VdomWeak::new(&inner).render().await?;
+
+        Ok(old)
     }
 
     /// Execute `f` with a reference to this virtual DOM's root rendering
     /// component.
     ///
     /// To ensure exclusive access to the root rendering component, the
     /// invocation takes place on a new tick of the micro-task queue.
-    pub fn with_component<F, T>(&self, f: F) -> impl Future<Item = T, Error = VdomDroppedError>
+    pub async fn with_component<F, T>(&self, f: F) -> Result<T, VdomDroppedError>
     where
         F: 'static + FnOnce(&mut dyn RootRender) -> T,
     {
-        futures::future::ok(self.inner.upgrade())
-            .and_then(|inner| inner.ok_or(()))
-            .map_err(|_| VdomDroppedError {})
-            .and_then(|inner| {
-                let mut f = Some(f);
-                let promise = js_sys::Promise::resolve(&JsValue::null());
-                JsFuture::from(promise)
-                    .map_err(|_| VdomDroppedError {})
-                    .map(move |_| {
-                        let f = f.take().unwrap_throw();
-                        let mut exclusive = inner.exclusive.borrow_mut();
-                        f(exclusive.component_raw_mut())
-                    })
-            })
```

Comment:
Unsure if the Promise and JsFuture are required, I ran the tests without and found no issues.